cpp/src/snmg/degree/degree.cuh

```diff
@@ -0,0 +1,166 @@
+#include "rmm_utils.h"
+
+namespace cugraph {
+  /**
+   * Single node multi-GPU method for degree calculation on a partitioned graph.
+   * @param x Indicates whether to compute in degree, out degree, or the sum of both.
+   *    0 = in + out degree
+   *    1 = in-degree
+   *    2 = out-degree
+   * @param part_off The vertex partitioning of the global graph
+   * @param off The offsets array of the local partition
+   * @param ind The indices array of the local partition
+   * @param degree Pointer to pointers to memory on each GPU for the result
+   * @return Error code
+   */
+  template<typename idx_t>
+  gdf_error snmg_degree(int x, size_t* part_off, idx_t* off, idx_t* ind, idx_t** degree) {
+    sync_all();
+    SNMGinfo env;
+    auto i = env.get_thread_num();
+    auto p = env.get_num_threads();
+
+    // Getting the global and local vertices and edges
+    size_t glob_v = part_off[p];
+    size_t loc_v = part_off[i + 1] - part_off[i];
+    idx_t tmp;
+    CUDA_TRY(cudaMemcpy(&tmp, &off[loc_v], sizeof(idx_t), cudaMemcpyDeviceToHost));
+    size_t loc_e = tmp;
+
+    // Allocating the local result array, and setting all entries to zero.
+    idx_t* local_result;
+    ALLOC_TRY((void** )&local_result, glob_v * sizeof(idx_t), nullptr);
+    thrust::fill(rmm::exec_policy(nullptr)->on(nullptr), local_result, local_result + glob_v, 0);
+
+    // In-degree
+    if (x == 1 || x == 0) {
+      dim3 nthreads, nblocks;
+      nthreads.x = min((idx_t) loc_e, (idx_t) CUDA_MAX_KERNEL_THREADS);
+      nthreads.y = 1;
+      nthreads.z = 1;
+      nblocks.x = min((idx_t) ((loc_e + nthreads.x - 1) / nthreads.x), (idx_t) CUDA_MAX_BLOCKS);
+      nblocks.y = 1;
+      nblocks.z = 1;
+      degree_coo<idx_t, idx_t> <<<nblocks, nthreads>>>((idx_t) loc_e,
+                                                       (idx_t) loc_e,
+                                                       ind,
+                                                       local_result);
+    }
+
+    // Out-degree
+    if (x == 2 || x == 0) {
+      dim3 nthreads, nblocks;
+      nthreads.x = min((idx_t) loc_v, (idx_t) CUDA_MAX_KERNEL_THREADS);
+      nthreads.y = 1;
+      nthreads.z = 1;
+      nblocks.x = min((idx_t) ((loc_v + nthreads.x - 1) / nthreads.x), (idx_t) CUDA_MAX_BLOCKS);
+      nblocks.y = 1;
+      nblocks.z = 1;
+      degree_offsets<idx_t, idx_t> <<<nblocks, nthreads>>>((idx_t) loc_v,
+                                                           (idx_t) loc_e,
+                                                           off,
+                                                           local_result + part_off[i]);
+    }
+
+    // Combining the local results into global results
+    sync_all();
+    treeReduce<idx_t, thrust::plus<idx_t> >(env, glob_v, local_result, degree);
+
+    // Broadcasting the global result to all GPUs
+    treeBroadcast(env, glob_v, local_result, degree);
+
+    return GDF_SUCCESS;
+  }
+
+  template<>
+  gdf_error snmg_degree<int64_t>(int x,
```

Comment:
Because atomicAdd() doesn't work with 64 bit integers, the specialization for int64_t calls degree_coo<int64_t, double> and then converts the results from double to int64_t with another kernel.

Reply:
got it